### DISCUSSION TITLE
Using RHVoice for TTS in Thorium instead of Windows Default

### DISCUSSION THREAD

**Discussion by brenojac** [edrlab/thorium-reader#2095] [General]
Dear Thorium Team and community,

We've been trying to setup the Leticia Voice for Brazillian Portuguese (https://louderpages.org/leticia) as a TTS option in Thorium. It is SAPI 5 compatible, according to them:  "You can also use RHVoice with other applications for Windows, e.g. JAWS, if they are compatible with Microsoft Speech API Version 5".

Now, we followed threads #1792, #1648, #1151, but were not able to make the Leticia voice show up in the list. We even tried a Windows Registry hack (https://stackoverflow.com/questions/47379725/how-do-i-add-a-voice-language-to-speechsynthesis), as suggested by @danielweck, but unfortunately it didn't solve it. I decided to open this issue so that perhaps we could investigate and find a solution together that may be useful to us and others in the future. 

Here are the steps we've taken this far: 

1) Browse https://rhvoice.org/languages/ and click on Portuguese. It will redirect to https://louderpages.org/leticia

2) Download and install https://rhvoice.org/download/RHVoice-Brazilian-Portuguese-voice-Leticia-F123-v4.6.1021.18-setup.exe

3) Try to list the new TTS voices in Thorium or Windows or even in the browser using https://jsbin.com/wuqumukesu/edit?js,console,output 

4) Unfortunately in either case, Leticia voice does not show up as an alternative anywhere. 

Further investigations have shown us that Windows TTS voices are stored at C:\Windows\Speech_OneCore\Engines\TTS\, while RHVoice Leticia Voice are at C:\ProgramData\Olga Yakovleva\RHVoice\ and apparently in a different format. Not sure how to go from here. 

One hint I suspect comes from RHVoice page stating that "you should configure your software to use RHVoice as its speech synthesizer", which suggest it as alternative to Windows TTS?  

Has anyone in the community gone through this problem? Suggestions / hints are welcome. 

Thanks in advance

Breno

**Comment by danielweck** [comment]
Thank you very much for your detailed analysis!
As this is not an issue in Thorium per-se, I am transferring to a GitHub discussion.

**Reply by brenojac** [reply]
Hello Daniel,

Thank you for the reply and sorry for posting as an issue. I was not aware you guys had a discussion forum here.

**Comment by danielweck** [comment]
I think the best next course of action is to create a reproducible test case: install the voice on a Windows machine (computer or virtual), create a simple HTML + Javascript test webpage that can be opened in Google Chrome or any other Chromium-based web browser (although Microsoft Edge *might* in fact behave differently as it is capable of interfacing with Windows proprietary or platform-specific APIs). The test webpage simply lists available TTS voices via the Web Speech API ... the test fails if the Leticia voice is not listed. That bug report can be sent to the owners / maintainers of this synthetic speech voice.

**Reply by brenojac** [reply]
Great. I did just that. I posted as an issue and they'll be investigating with me. I'll update this thread as soon as I get new information.

**Comment by zstanecic** [comment]
Hi to all!
How do you list tts voices in thorium? I am using it with NVDA to debug the problem with Leticia.
Zvonimir stanecic, RHvoice maintainer for some languages

**Reply by danielweck** [reply]
Thorium discovers TTS voices via the Web Speech API through Javascript in the Chromium web browser engine. Same as the script posted above.

**Reply by danielweck** [reply]
window.speechSynthesis.getVoices();

**Comment by zstanecic** [comment]
no way. we found the problem.
unfortunatelly, Thorium doesn't recognize tts voices via the sapi interface, just the online one from microsoft.

**Reply by danielweck** [reply]
that's weird, it's normally the exact opposite :) the WebSpeech API picks from SAPi5, excluding OneCore premium TTS voices ?!

**Comment by zstanecic** [comment]
I am trying to use the version of Thorium from Microsoft store. 

But, I don’t see options for configuration of Microsoft voices via thorium.

**Reply by brenojac** [reply]
Dear @danielweck @zstanecic,

Let me add to the discussion, given my investigations this far. Thorium is able to list windows TTS voice which are locally installed and listed at the folder C:\Windows\Speech_OneCore\Engines\TTS. I can add and remove new voices using Windows voice configuration interface: 

![image](https://github.com/edrlab/thorium-reader/assets/563016/a6af2c1d-9b45-4932-97be-35debd6dea3d)

@zstanecic, After adding any voices in this fashion, it will get listed in Thorium when you open an EPUB or other format and click on the Activate TTS button: 
![image](https://github.com/edrlab/thorium-reader/assets/563016/81ec8f13-0e26-4562-a4e0-d446e48c2928)
And 
![image](https://github.com/edrlab/thorium-reader/assets/563016/139defa4-d3cb-4d3f-8158-692ea0e5edaa)

I'm not sure if you tried like that. Now, when installing Leticia voice, the files wont go to the Windows TTS folder, which is where voices are listed in Thorium (sorry if I'm mistaken about this, but that's how I'm understanding what Thorium is doing). Everything installed by Leticia voice goes to  C:\ProgramData\Olga Yakovleva\RHVoice\, am I correct? 

Now, in theory the SAPI 5 interfaces would be able to retrieve voices from both Windows TTS and Leticia's RHVoice, given they are compatible? That still needs to be understood in my opinion. When I list the voices using https://jsbin.com/wuqumukesu/edit?js,console,output, I still get the same listing as Thorium. Only Windows TTS voices. If this listing goes through SAPI, why cant it locate Leticia's voice as well? 

![image](https://github.com/edrlab/thorium-reader/assets/563016/d820925b-7985-46f2-bc91-9b5abb856fbf)

Breno

**Comment by zstanecic** [comment]
Hi,

This way you describe here shouldn’t be used at all.

Your screenshots are not helping me, as I am totally blind.

Note that Sapi voices should be looked up from the registry.

**Comment by brenojac** [comment]
Dear @zstanecic 

Sorry my last post was not helpful. Have you managed to test the TTS feature of Thorium? It is not available as a configuration option. Instead, the available voices show up when clicking in the Activate TTS Button, while reading an ebook. On the top center there'll be a list of voices to select. But the voices listed are the ones installed as a Windows  voice package. 

You mentioned the voices are retrieved from the Windows Registry. Could you point out where Leticia voice is located? That may help us find out why Thorium is unable to find it. 

Breno

**Comment by zstanecic** [comment]
Windows voice package? So, it means only Narrator natural voices or OneCore voices.

RHVoice stores his info as follows:

32-bit registry key: Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\SPEECH\Voices\TokenEnums\RHVoice

64-bit: Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\SPEECH\Voices\TokenEnums\RHVoice

I don’t have a way to choose voices when I press the activate text to speech button.

I don’t have combobox or any other element.

**Reply by brenojac** [reply]
@zstanecic, there's actually a combobox where voices are listed. I installed NVDA and it was able to read them, but only when I point the mouse to it. The first item on the list is Voice: system default. When clicking on the combobox, it lists the rest of the voices available. 

@danielweck Could you please help us figure out why Thorium is not able to find Leticia's voice? Entries are being add in the windows registry as described above. What else could you suggest to make Thorium list Leticia's voice?

Thanks in advance.

Breno

**Reply by danielweck** [reply]
Thank you both for your investigations. Before we focus on Thorium itself, could you please confirm whether or not the TTS voice is listed in Chromium, or Google Chrome. Please check Microsoft Edge as well, just to see if there are differences. Please use the Web Speech API JavaScript that was previously mentioned. If the TTS voice appears successfully, then Thorium should find it too. Conversely, if the TTS voice is missing, then this explains why Thorium doesn't see it. Fixing the problem for Chromium will solve it in Thorium too.

**Comment by zstanecic** [comment]
I don’t see this combobox with a keyboard, so it is inaccessible. 
Thanks!

**Comment by zstanecic** [comment]
Hi,

With script provided above, I cannot find anoy of the RHVoice voices which were locally installed.

Thanks,

Zvonimir Stanecic



-- 
Ta wiadomość e-mail została sprawdzona pod kątem wirusów przez oprogramowanie antywirusowe Avast.
www.avast.com

**Reply by brenojac** [reply]
Hello,

I have tested the script provided using the Edge browser just to make sure that Leiticia voice is not listed at all. It lists all installed windows voices (narrator natural voices / Onecore), but Leticia's voice is not there. This javascript in the browser implements the same logic to find voices that Thorium does. 

Thus, I think we can safely state that Leticia voice, once installed, is not being exposed by SAPI 5. @zstanecic I could see in some RHVoice releases that the Olga developer was responsible for making RHVoices compatible with the Speech API. Could she, perhaps, help us figure this out? 

We're determined to find a solution to this problem as Leticia's voice is considered to be the best one to read our interactive ebooks. 

Thanks again for all your attention and effort into solving this. 

Breno

**Comment by brenojac** [comment]
Dear @danielweck @zstanecic ,

Hope your both well. We keep investigating how to use Letícia's voice from Thorium Reader. 

Using NVDA with Leticia requires an additional plug-in to make it work. We're willing to develop one for Thorium in case it is also necessary. We are really eager to use Letícia voice as we're sure she's the best voice today for brazilian portuguese. 

@zstanecic could we be in closer contact with RHVoice team to make this happen? We have a dedicated developer who could try to go deeper into investigating the use of RHVoices in Thorium either through SAPI or through an additional plug-in. 

Finally, is there any additional documentation we could use to understand RHVoice's are exposed to SAPI 5?

Best regards,

- Breno

**Comment by HadrienGardeur** [comment]
@brenojac have you tested Microsoft Natural voices voices for Portuguese as well?

On Edge, Microsoft now provides:

- three voices for Brazilian Portuguese: Thalita, Antonio and Francisca
- two voices for Portuguese from Portugal: Duarte and Raquel

In my experience with other languages, Microsoft Natural voices are some of the best available on the market and the best ones that come preloaded in a browser or an OS.

On Windows 11, the selection is smaller currently but it includes Antonio and Francisca with the most recent Windows 11 updates: https://support.microsoft.com/en-us/windows/appendix-a-supported-languages-and-voices-4486e345-7730-53da-fcfe-55cc64300f01
This support page doesn't indicate that these are Brazilian Portuguese voices, but in Edge they're both identified by `pt-br` as their language tag.

As a side note, I've been documenting high quality voices across platforms in a documentation repo at https://github.com/HadrienGardeur/web-speech-recommended-voices

Portuguese is fairly high in my to-do list, I've opened an issue to handle it next along with German: https://github.com/HadrienGardeur/web-speech-recommended-voices/issues/6

**Reply by brenojac** [reply]
Dear @HadrienGardeur,

Thank you for your feedback. I have managed to install these natural voices in a Windows 11 box. Their quality is really great!

Unfortunately, they wont get listed in Thorium Reader as alternative. Have you managed to use them in Thorium for TTS functionality?

@danielweck Do you think Microsoft Natural voices should be listed via SAPI5 just like any other voice?

Breno

**Reply by danielweck** [reply]
Once again, what I can state at his juncture is that if the TTS voices are not discovered via the Web Speech API by "vanilla" Chromium-based web browsers like Google Chrome or Microsoft Edge, then this explains why Thorium is not listing them either. The Chromium webview component embedded in Thorium is slightly older than the one shipped in the most recent Google Chrome, but not far behind as it is provided by a recent Electron release.

There is one possible escape hatch to gain access to proprietary / protected synthetic voices: break out of the Web Speech Javascript API, instead rely on network requests (authenticated cloud service) or use some kind of native platform-specific bridge. All technically possible but not on the development roadmap due to resource / feature prioritization.

My recommendation at this stage is to push for SAPI5 compliant TTS voices to be supported in the Web Speech API ... report bugs in the Chromium project if all else fails.

**Reply by HadrienGardeur** [reply]
@danielweck I agree that we should push for this support at a Chromium level and that's exactly what I plan on doing in the next few days by either identifying existing issues in their tracker (and voting for them) or opening new ones.

I'll track everything using a label at https://github.com/HadrienGardeur/web-speech-recommended-voices/issues?q=is%3Aissue+is%3Aopen+label%3Aexternal-issue

**Reply by HadrienGardeur** [reply]
On a related note: Google also offers high quality voices in Chrome that are not visible in other Chromium based projects. 

This is more of a stretch since these are online voices with server-side costs incurred by Google, but having them in Chromium and therefore Thorium would be a very nice addition.

**Reply by brenojac** [reply]
Hello @danielweck @HadrienGardeur 

Thanks a lot for the feedback. Hadrien, please, let us know if we can be of any help making those voices available.

It turns out that according to our blind users, they will use their preferred screen readers, such as NVDA or JAWS, to navigate and also to read aloud the interactive ebooks. In other words, they probably wont use Thorium TTS functionality at all. Yet, I still think it is valid effort to make voices such as Leticia's from RHVoice or Microsoft Natural Voices available as a better TTS alternative. 

Breno

**Reply by HadrienGardeur** [reply]
In my experience, read aloud targets a very different audience than screen readers and it's important to make sure that both options are available.

People with dyslexia for example won't use a screen reader, but could greatly benefit from having a read aloud feature available alongside with the ability to customize the reading experience to their liking.

These high quality voices vastly improve the experience, but it's a shame that voice selection is in such a dire state.

**Reply by brenojac** [reply]
Hello @HadrienGardeur ,

You made a very relevant point on the usefulness of the TTS feature. That made us even eager to make this work. One relevant point for us is the ability to use those high quality voices offline. Most schools in Brazil lack a stable infrastructure and thus reading offline is a must-have for us. 

Considering both Google and Microsoft Voices are, to the best of our knowledge, cloud-dependent, and also possibly not cross-platform (windows only?), I can only see Leticia from RHVoice as a proper candidate for offline use. It supports Linux, Windows, Android and MacOS. 

What do you think? We're willing to make an effort to list Leticia and other voices from the RHVoice available from Thorium / Chromium. They have an NVDA plugin where it gets listed like this: 

![image](https://github.com/edrlab/thorium-reader/assets/563016/a1968b35-9aef-4b0d-ac82-0d9bf74223f9)

Now, that looks simple enough for a user. NVDA is a desktop app and is still using Microsoft Speech API to list RHVoice's Leticia. How come we cant do that from Chromium / Thorium? 

Breno

**Reply by danielweck** [reply]
hello Breno, I posted a message in the RHVoice repository: https://github.com/RHVoice/RHVoice/issues/862#issuecomment-2088465808

**Reply by HadrienGardeur** [reply]
@brenojac as I've pointed out before, I'm documenting voices across platforms/browsers right now. Here's the JSON that I've created for Portuguese for instance, which covers both Brazil and Portugal: https://github.com/HadrienGardeur/web-speech-recommended-voices/blob/main/json/pt.json

Based on what I've tested so far, there are a number of higher quality voices available offline. If I had to rank them:

- Microsoft Natural Voices for Windows 11 (Antonio and Francisca)
- Chrome OS Natural Voices and Android voices (three voices for Brazil)
- and finally Apple voices on macOS, iOS and iPadOS (Fernanda, Luciana and Felipe)

It's unfortunate that due to various issues, some of these voices are not available through the WebSpeech API but that's something that I'd like to address [by reporting these bugs](https://github.com/users/HadrienGardeur/projects/2/views/1) and working together with international organizations to shake things up.

With the current expectations that all desktop and mobile platforms will be pushing for dedicated AI hardware this year, this will only get better and better over time. Unlike AI chatbots, ML-based voices are already widely useful today and will be a much easier problem to "solve".

**Reply by HadrienGardeur** [reply]
For more technical details about what Microsoft is working on, I found the following article yesterday: https://techcommunity.microsoft.com/t5/ai-azure-ai-services-blog/azure-neural-tts-now-available-on-devices-for-disconnected-and/ba-p/3716797

Based on what I can see, most voices are made available through Azure first before they become available in Edge and finally Windows 11 in offline mode.
Azure's voice galley is a good way to peek into the near future of voices on Windows: https://speech.microsoft.com/portal/voicegallery

**Reply by brenojac** [reply]
Dear @HadrienGardeur and @danielweck,

Sorry for the late answer. Hadrien, the work you're doing is really significant for us. The voices you listed above have been tested by some of our users (Antonio and Francesca) and they are truly of high quality. Being available offline is paramount. We cant wait to see them available via Thorium. 

We'll take a deeper look when we get back from Paris. We (NEES research group/UFAL) and some members of the Brazilian government will be attending DPS this year. Also, we'll have the chance to meet Laurent and hopefully you two, as well? If that's the case, we could discuss the current state of these natural voices in the upcoming versions of Thorium. Please count on us to make this happen.  

Hope to see you soon! 

- Breno

**Reply by HadrienGardeur** [reply]
Hello @brenojac,

I'm glad that this work is useful. I'll be at DPS this year as well and should also attend your meeting with Laurent.

Since this last message, I've made some progress when it comes to voice selection with:

- Best practices: https://github.com/HadrienGardeur/read-aloud-best-practices/blob/main/voice-selection.md
- Known-issues: https://github.com/users/HadrienGardeur/projects/2/views/1?sliceBy%5Bvalue%5D=Voice+selection
- and further languages supported in the recommended voices project: https://github.com/HadrienGardeur/web-speech-recommended-voices

I've also created a fairly basic demo to test recommended voices: https://hadriengardeur.github.io/web-speech-recommended-voices/demo/

Ideally, Thorium should follow these best practices and automatically default to a Brazilian voice if the content is in Portuguese and the device is set to Brazilian Portuguese (`pt-BR`).

Using the recommended voices that I've documented, the following voices would be used on Windows (ordered by priority):
- Francisca
- Thalita (not available yet in Windows 11, but pre-loaded in Edge)
- Antonio
- Maria
- Daniel

If the device is not set to `pt-BR` or if none of the Brazilian voices are available, it would fallback to voices from Portugal in the following order:

- Raquel
- Duarte
- Helia

On macOS:

- Fernanda
- Luciana (pre-loaded by macOS)
- Felipe

With a similar fallback to voices from Portugal, if needed:

- Catarina
- Joana (pre-loaded by macOS)
- Joaquim

Due to the lack of support for Microsoft Natural voices through the `getVoices()` API, I haven't been able to document them yet.

@gexgd0419 and @Leroy-X how are these voices identified (`name` and `voiceURI` returned by `getVoices()`) if you patch Chrome for Windows?

**Reply by Leroy-X** [reply]
> * Thalita (not available yet in Windows 11, but pre-loaded in Edge)

Hi, NaturalVoiceSAPIAdapter can simulate voices in Edge into SAPI5, including natural voices from Narrator. After modifying the registry, these voices were automatically detected.

**Reply by gexgd0419** [reply]
Narrator natural voices and online natural voices in Microsoft Edge are not freely available to third-party developers. Developers wanting to use the natural voices are expected to either pay for the online version, or go through some [limited access review](https://learn.microsoft.com/en-us/legal/cognitive-services/speech-service/embedded-speech/limited-access-embedded-speech?context=/azure/ai-services/speech-service/context/context) to get the offline version.

[NaturalVoiceSAPIAdapter](https://github.com/gexgd0419/NaturalVoiceSAPIAdapter), which can make the natural voices mentioned above accessible as SAPI 5 voices, is basically a hack that uses some encryption keys to gain access to those voices. It will be broken once Microsoft decides to change the keys, but it works at least for now.

> how are these voices identified (`name` and `voiceURI` returned by getVoices()) if you patch Chrome for Windows?

In Chrome and Chromium Edge, the `voiceURI` of each voice is the same as its `name`, e.g. `Microsoft Aria Online (Natural) - English (United States)`.

In Firefox, the `voiceURI` is like this: `urn:moz-tts:sapi:Microsoft Aria Online (Natural) - English (United States)?en-US`.

Firefox supports all SAPI 5 voices out of the box, including NaturalVoiceSAPIAdapter voices. Their `voiceURI` begin with `urn:moz-tts:sapi:`.

Chromium browsers only have a somewhat broken support for OneCore voices on Windows 10/11. But if you copy `HKLM\Software\Microsoft\Speech\Voices` to `HKLM\Software\Microsoft\Speech_OneCore\Voices`, Chromium browsers can still use those SAPI 5 voices, because they *are* using SAPI 5 under the hood.

**Reply by HadrienGardeur** [reply]
> Narrator natural voices and online natural voices in Microsoft Edge are not freely available to third-party developers. Developers wanting to use the natural voices are expected to either pay for the online version, or go through some [limited access review](https://learn.microsoft.com/en-us/legal/cognitive-services/speech-service/embedded-speech/limited-access-embedded-speech?context=/azure/ai-services/speech-service/context/context) to get the offline version.

Thanks, that's useful information. I understand the tension on the business model side of things: Microsoft wants developers to use Azure Cloud to get access to these voices. But considering the importance of TTS for accessibility use cases and the fact that some of these voices can run locally, I think they're making the wrong bet.

Chrome OS has some of the least user-friendly voice names that I've seen but you always get access to:

- medium quality local voices
- higher quality online voices
- and for a few languages now, you can also install higher quality local voices (they're also calling them "Natural voices")

I really think that's the right approach and something that Microsoft will hopefully consider as well.

Given the nature of what Thorium does, I think it would probably qualify for a limited access review, but I wonder if that's purely for native apps (Thorium is built with Electron).

**Reply by gexgd0419** [reply]
> Given the nature of what Thorium does, I think it would probably qualify for a limited access review

In the [limited access review](https://learn.microsoft.com/en-us/legal/cognitive-services/speech-service/embedded-speech/limited-access-embedded-speech?context=/azure/ai-services/speech-service/context/context) page:

> As a Limited Access feature, embedded Speech requires registration. Only customers managed by Microsoft, meaning those who are working directly with Microsoft account teams, are eligible for access.

> The embedded Speech is made available to customers under the terms governing their subscription to Microsoft Azure Services

Also in the application form:

> Access is limited to Microsoft commercial customers that meet the following requirements: 
>
> a. Valid Azure subscription ID - Applications may not be submitted on behalf of other organizations and these products may not be deployed to any environment not owned by the applicant organization (your employer). 
>
> b. Valid business scenario/use case - Azure Speech is a cloud first product. The use of embedded speech is only approved in cases where our cloud or container offerings cannot meet the needs, for instance:
>
>   •    Environment or device(s) with periods of zero connectivity to internet. 
>   •    Organization under strict regulation of not sending any kind of data back to cloud.  
>
> c. Application completed as instructed - Please pay close attention to guidance provided throughout the application to ensure you provide all the necessary information required for approval.

**Reply by danielweck** [reply]
Re. Azure services: yes indeed, see for example how the DAISY Pipeline interfaces via a registered account (API key):
* https://github.com/daisy/pipeline/tree/master/modules/tts/tts-adapters/tts-adapter-azure
* https://github.com/daisy/pipeline-ui/blob/9d002032f485bf3ed27fca6aee6039d1ff0c8ff4/src/renderer/components/TtsEnginesConfig/index.tsx#L16-L24

**Reply by HadrienGardeur** [reply]
Microsoft is currently making a huge push against Apple with the release of Copilot+ PCs that come with dedicated hardware and new AI features. This would also be a nice showcase of what's possible on their platform.

I would argue that having these voices available offline is already a good argument for buying a Windows device rather than a macOS one, but that's probably not something that the Azure team cares about as much.

**Comment by zstanecic** [comment]
Hi,

Unfortunatelly no.

They use different protocols



-- 
Ta wiadomość e-mail została sprawdzona pod kątem wirusów przez oprogramowanie antywirusowe Avast.
www.avast.com

**Comment by Leroy-X** [comment]
Hi there, I found [NaturalVoiceSAPIAdapter](https://github.com/gexgd0419/NaturalVoiceSAPIAdapter) which can simulate natural voices as SAPI5. Unfortunately, I don't know how to add it to Thorium. If there are any solutions, please let me know. Thanks a lot.

**Reply by danielweck** [reply]
Open Google Chrome web browser, open the "web inspector" (developer console), and type `console.log(window.speechSynthesis.getVoices())`. Try the same with Microsoft Edge web browser. See if the TTS voice you need is included. If it is, then check again with Thorium.

**Reply by danielweck** [reply]
On MacOS I see 178 voices ... so it takes some time to manually check the names to find a particular TTS voice. You can accelerate this search process by filtering voices by keyword. Let us know if you need help write a Javascript filtering function.

<img width="1792" alt="Screenshot 2024-05-27 at 10 04 01" src="https://github.com/edrlab/thorium-reader/assets/625201/63b2db05-0512-4785-9890-3971df9f5854">

**Reply by Leroy-X** [reply]
Hi, I've checked in Chrome, and I can't get SAPI5 voice to work. As far as I know, Firefox supports SAPI5.

**Reply by danielweck** [reply]
SAPI5 voices are definitely supported by the Chrome / Chromium implementation of Javascript Web Speech API. However not all SAPI5 voices are exposed because of limitations / constraints at the level of the Windows operating system (which is why some registry hacks are necessary with some TTS voices)

**Reply by Leroy-X** [reply]
Hi, What should I do? It still doesn’t show up in the console. but it's works in balabolka.

**Comment by zstanecic** [comment]
Hi Lero!

Console newer is showing the sapi voices.

There we have a bug also with RHVoice.

Cromium is cloud-centric, this our local voices are not suitable for use 
with thorium.

**Reply by Leroy-X** [reply]
We seem to be talking about different things. I'm using the NaturalVoiceSAPIAdapter to simulate the natural voice of the Windows 11 narrator as SAPI5. Yes, they are completely local and offline. But they don't appear in the Chrome‘s developer console, I don't know how to add them. I saw someone say they succeeded by the regeditor. So I'm looking for help. :P thanks.

**Reply by danielweck** [reply]
Just to clarify once again: Chromium (the web browser engine) is not "cloud-centric" when it comes to TTS voices. As demonstrated in my screenshot, there are 178 local synthetic speech voices available via the Web Speech API on my MacOS computer. In Windows that number is smaller as the default selection of TTS voices shipped with the OS is limited (users can install additional language packs).
The fact that some SAPI5 voices do not appear in the standard Web Speech API is usually the operating system's fault. Some registry hacks can be found via searching on the web. We cannot officially recommend this because it is by definition a hack / workaround, but I can point you in that direction :)

**Reply by Leroy-X** [reply]
Thank you very much for your reply, I may have been disturbed by the information, I will go find something related to the registry :P

**Comment by gexgd0419** [comment]
So this is actually Chromium's fault...

[This commit](https://source.chromium.org/chromium/chromium/src/+/5411d7bd073a64b5f6b7c98d038744b7a1061d19) in Chromium repository added the ability to list OneCore voices.

Here's the code in question:

```c++
bool TtsPlatformImplBackgroundWorker::GetVoiceTokens(
    Microsoft::WRL::ComPtr<IEnumSpObjectTokens>* out_tokens) {
  if (S_OK ==
      SpEnumTokens(kSPCategoryOneCoreVoices, NULL, NULL, &(*out_tokens))) {
  } else if (S_OK != SpEnumTokens(SPCAT_VOICES, NULL, NULL, &(*out_tokens))) {
    return false;
  }
  return true;
}
```

You can see that once OneCore voices (`kSPCategoryOneCoreVoices`) are enumerated successfully, SAPI 5 voices (`SPCAT_VOICES`) will be ignored.

The committer thought that the OneCore voice list would include all SAPI voices as well (e.g. the OneCore voice `Microsoft Zira` and the SAPI voice `Microsoft Zira Desktop` are the same). However, users *can* install third-party SAPI voices. So the more proper way to do this would be querying both voice lists and merge them into one.

There is a [PR](https://chromium-review.googlesource.com/c/chromium/src/+/2971050) that tries to fix this, but it hasn't been merged yet, because the submitter didn't reply to any comment from the reviewer.

Copying the Speech registry key to Speech_OneCore works, because Chromium is actually using SAPI 5 to load OneCore voices, instead of using the appropriate WinRT API. It just tells the SAPI 5 system to load the voices at the OneCore voice registry key, which isn't correct, and some OneCore voices can fail to work correctly because of this.

**Reply by danielweck** [reply]
Brilliant investigation, thank you :)

**Comment by Leroy-X** [comment]
Thanks so much, now it works :DDDD

**Comment by danielweck** [comment]
Same problem with Nuance Vocalizer Expressive TTS 5.4
https://github.com/edrlab/thorium-reader/issues/2755

@Svetloslav please see this hack/workaround https://stackoverflow.com/questions/47379725/how-do-i-add-a-voice-language-to-speechsynthesis

**Comment by danielweck** [comment]
interesting findings: https://github.com/edrlab/thorium-reader/discussions/2945#discussioncomment-12923739

(Microsoft Narrator settings might be making changes to the Windows Registry in order to configure TTS Voices globally via SAPI5?)
